### PR TITLE
[Serializer] Add more parameter types

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #32179
| License       | MIT
| Doc PR        | N/A

We missed quite a lot in the Serializer, so I decided to open a separate PR instead of merging the changes into #33154.

### CODE REVIEW

**Comment by derrabus** [comment]
I'm already in love with fabbot's new powers. /cc @tigitz

Bildschirmfoto 2019-08-15 um 12 37 40

File: src/Symfony/Component/Serializer/Mapping/Factory/ClassResolverTrait.php [hunk 1]
```diff
@@ -25,7 +25,7 @@ trait ClassResolverTrait
     /**
      * Gets a class name for a given class or instance.
      *
-     * @throws InvalidArgumentException If the class does not exists
+     * @param object|string $value
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Mapping/Factory/ClassResolverTrait.php]: for a lower branch

File: src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php [hunk 2]
```diff
@@ -17,14 +17,13 @@ trait ObjectToPopulateTrait
      * Extract the `object_to_populate` field from the context if it exists
      * and is an instance of the provided $class.
      *
-     * @param string $class The class the object should be
-     * @param $context The denormalization context
-     * @param string $key They in which to look for the object to populate.
-     *                    Keeps backwards compatibility with `AbstractNormalizer`.
+     * @param string      $class The class the object should be
+     * @param string|null $key   They in which to look for the object to populate.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php]: lower branch

File: src/Symfony/Component/Serializer/Serializer.php [hunk 3]
```diff
@@ -109,7 +109,7 @@ public function __construct(array $normalizers = [], array $encoders = [])
     /**
      * {@inheritdoc}
      */
-    final public function serialize($data, $format, array $context = [])
+    public function serialize($data, string $format, array $context = [])
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Serializer.php]: no pb on my side, but why removing `final`?

File: src/Symfony/Component/Serializer/Serializer.php [hunk 4]
```diff
@@ -175,31 +175,31 @@ public function normalize($data, $format = null, array $context = [])
      *
      * @throws NotNormalizableValueException
      */
-    public function denormalize($data, $type, $format = null, array $context = [])
+    public function denormalize($data, string $class, string $format = null, array $context = [])
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Serializer.php]: renaming for a lower branch if confirmed

File: src/Symfony/Component/Serializer/Tests/Normalizer/AbstractObjectNormalizerTest.php [hunk 5]
```diff
@@ -289,66 +289,58 @@ class SerializerCollectionDummy implements SerializerInterface, DenormalizerInte
...
     {
     }
 
-    public function deserialize($data, $type, $format, array $context = [])
+    public function deserialize($data, string $type, string $format, array $context = [])
     {
     }
 
-    public function denormalize($data, $type, $format = null, array $context = [])
+    public function denormalize($data, string $class, string $format = null, array $context = [])
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/AbstractObjectNormalizerTest.php]: renaming to be confirmed, and applied to a lower branch

File: src/Symfony/Component/Serializer/Tests/Normalizer/AbstractObjectNormalizerTest.php [hunk 6]
```diff
@@ -289,66 +289,58 @@ class SerializerCollectionDummy implements SerializerInterface, DenormalizerInte
...
     }
 
-    protected function isAllowedAttribute($classOrObject, $attribute, $format = null, array $context = [])
+    protected function isAllowedAttribute($classOrObject, string $attribute, string $format = null, array $context = [])
     {
         return true;
     }
 
-    public function instantiateObject(array &$data, $class, array &$context, \ReflectionClass $reflectionClass, $allowedAttributes, string $format = null)
+    protected function instantiateObject(array &$data, string $class, array &$context, \ReflectionClass $reflectionClass, $allowedAttributes, string $format = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/AbstractObjectNormalizerTest.php]: lower branch?

File: src/Symfony/Component/Serializer/Tests/Normalizer/AbstractObjectNormalizerTest.php [hunk 7]
```diff
@@ -289,66 +289,58 @@ class SerializerCollectionDummy implements SerializerInterface, DenormalizerInte
...
     {
         return parent::instantiateObject($data, $class, $context, $reflectionClass, $allowedAttributes, $format);
     }
-
-    public function serialize($data, $format, array $context = [])
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/AbstractObjectNormalizerTest.php]: lower branch?

**Review by nicolas-grekas — COMMENTED** [review]
sweet :)

File: src/Symfony/Component/Serializer/Serializer.php (hunk 3 above)
Comment (derrabus) [reply src/Symfony/Component/Serializer/Serializer.php]: That was a C&P fail. Reverted.

File: src/Symfony/Component/Serializer/Serializer.php (hunk 4 above)
Comment (derrabus) [reply src/Symfony/Component/Serializer/Serializer.php]: I've renamed the parameter in order to match the interface.

**Review by dunglas — APPROVED** [review]

**Review by Tobion — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Can be rebased

**Comment by Tobion** [comment]
Thank you @derrabus.

**Pull request merged**